File: src/app/api/routes/memory.py
"""
Memory management API endpoints with comprehensive logging and telemetry.

Provides REST endpoints for managing user conversation memory including:
- Conversation history retrieval
- Message search functionality
- Memory statistics and analytics
- Memory cleanup and management
"""

from __future__ import annotations

import logging
from typing import Annotated, Any

from fastapi import APIRouter, Depends, HTTPException, Query, Request
from fastapi.responses import JSONResponse
from opentelemetry import trace
from opentelemetry.trace import Status, StatusCode
from pydantic import BaseModel, Field

from app.api.routes.chat import get_optional_user, get_user_email
from app.api.services.memory_service import get_memory_service
from app.observability.app_insights import app_insights

router = APIRouter(prefix="/memory", tags=["memory"])
logger = logging.getLogger(__name__)
tracer = trace.get_tracer(__name__)


class ConversationHistoryRequest(BaseModel):
    """Request model for retrieving conversation history."""

    limit: int | None = Field(
        default=20, ge=1, le=100, description="Maximum number of messages to retrieve"
    )
    thread_id: str | None = Field(default=None, description="Filter by specific thread ID")
    include_metadata: bool = Field(
        default=False, description="Include message metadata in response"
    )


class MessageSearchRequest(BaseModel):
    """Request model for searching messages."""

    query: str = Field(min_length=2, max_length=200, description="Search query text")
    limit: int = Field(default=10, ge=1, le=50, description="Maximum number of results to return")
    thread_id: str | None = Field(default=None, description="Filter by specific thread ID")


class ConversationHistoryResponse(BaseModel):
    """Response model for conversation history."""

    user_id: str
    messages: list[dict[str, Any]]
    total_count: int
    thread_id: str | None
    has_more: bool


class MessageSearchResponse(BaseModel):
    """Response model for message search."""

    user_id: str
    query: str
    results: list[dict[str, Any]]
    total_found: int


class UserMemoryStatsResponse(BaseModel):
    """Response model for user memory statistics."""

    user_id: str
    total_messages: int
    recent_user_messages: int
    recent_assistant_messages: int
    memory_utilization_percent: float
    max_memory_limit: int


@router.get("/history", response_model=ConversationHistoryResponse)
async def get_conversation_history(
    request: Request,
    current_user: Annotated[dict[str, Any], Depends(get_optional_user)],
    limit: int = Query(default=20, ge=1, le=100, description="Maximum messages to return"),
    thread_id: str | None = Query(default=None, description="Filter by thread ID"),
    include_metadata: bool = Query(default=False, description="Include message metadata"),
) -> ConversationHistoryResponse:
    """
    Retrieve user's conversation history with intelligent filtering and pagination.

    This endpoint provides access to the user's stored conversation history,
    enabling context-aware AI interactions across sessions.
    """
    with tracer.start_as_current_span("api.memory.history") as span:
        user_email = get_user_email(current_user)
        span.set_attribute("user_id", user_email)
        span.set_attribute("limit", limit)
        span.set_attribute("has_thread_filter", thread_id is not None)
        span.set_attribute("include_metadata", include_metadata)

        try:
            memory_service = get_memory_service()

            # Get conversation history
            messages = await memory_service.get_user_conversation_history(
                user_id=user_email,
                limit=limit,
                thread_id=thread_id,
                include_metadata=include_metadata,
            )

            # Get total message count for pagination info
            stats = await memory_service.get_user_memory_stats(user_email)
            total_count = stats["total_messages"]

            response = ConversationHistoryResponse(
                user_id=user_email,
                messages=messages,
                total_count=total_count,
                thread_id=thread_id,
                has_more=total_count > limit,
            )

            span.set_attribute("messages_returned", len(messages))
            span.set_attribute("total_messages", total_count)

            logger.info(
                "Conversation history retrieved successfully",
                extra={
                    "user_id": user_email,
                    "messages_returned": len(messages),
                    "total_messages": total_count,
                    "thread_id": thread_id,
                },
            )

            # Track API usage in Application Insights
            app_insights.track_custom_event(
                "ConversationHistoryAPICall",
                {
                    "user_id": user_email,
                    "messages_returned": len(messages),
                    "total_messages": total_count,
                    "has_thread_filter": thread_id is not None,
                },
            )

            return response

        except Exception as exc:
            span.record_exception(exc)
            span.set_status(Status(StatusCode.ERROR, str(exc)))

            logger.error(
                "Failed to retrieve conversation history",
                extra={"user_id": user_email, "error": str(exc), "error_type": type(exc).__name__},
                exc_info=True,
            )

            app_insights.track_exception(exc)

            raise HTTPException(
                status_code=500, detail="Failed to retrieve conversation history"
            ) from exc


@router.post("/search", response_model=MessageSearchResponse)
async def search_messages(
    search_request: MessageSearchRequest,
    current_user: Annotated[dict[str, Any], Depends(get_optional_user)],
) -> MessageSearchResponse:
    """
    Search through user's message history using full-text search.

    This endpoint enables users to find specific conversations or topics
    from their interaction history, improving the AI's ability to reference
    past discussions and maintain context.
    """
    with tracer.start_as_current_span("api.memory.search") as span:
        user_email = get_user_email(current_user)
        span.set_attribute("user_id", user_email)
        span.set_attribute("search_query", search_request.query)
        span.set_attribute("limit", search_request.limit)

        try:
            memory_service = get_memory_service()

            # Perform message search
            results = await memory_service.search_user_messages(
                user_id=user_email,
                search_query=search_request.query,
                limit=search_request.limit,
                thread_id=search_request.thread_id,
            )

            response = MessageSearchResponse(
                user_id=user_email,
                query=search_request.query,
                results=results,
                total_found=len(results),
            )

            span.set_attribute("results_count", len(results))

            logger.info(
                "Message search completed successfully",
                extra={
                    "user_id": user_email,
                    "search_query": search_request.query,
                    "results_count": len(results),
                    "thread_id": search_request.thread_id,
                },
            )

            # Track search API usage
            app_insights.track_custom_event(
                "MessageSearchAPICall",
                {
                    "user_id": user_email,
                    "results_count": len(results),
                    "query_length": len(search_request.query),
                    "has_thread_filter": search_request.thread_id is not None,
                },
            )

            return response

        except Exception as exc:
            span.record_exception(exc)
            span.set_status(Status(StatusCode.ERROR, str(exc)))

            logger.error(
                "Message search failed",
                extra={
                    "user_id": user_email,
                    "search_query": search_request.query,
                    "error": str(exc),
                    "error_type": type(exc).__name__,
                },
                exc_info=True,
            )

            app_insights.track_exception(exc)

            raise HTTPException(status_code=500, detail="Message search failed") from exc


@router.get("/stats", response_model=UserMemoryStatsResponse)
async def get_memory_stats(
    current_user: Annotated[dict[str, Any], Depends(get_optional_user)],
) -> UserMemoryStatsResponse:
    """
    Get comprehensive statistics about user's memory usage and patterns.

    Provides insights into conversation patterns, memory utilization,
    and analytics that can help improve the AI assistant's performance.
    """
    with tracer.start_as_current_span("api.memory.stats") as span:
        user_email = get_user_email(current_user)
        span.set_attribute("user_id", user_email)

        try:
            memory_service = get_memory_service()

            # Get comprehensive memory statistics
            stats = await memory_service.get_user_memory_stats(user_email)

            response = UserMemoryStatsResponse(**stats)

            span.set_attribute("total_messages", stats["total_messages"])
            span.set_attribute("memory_utilization", stats["memory_utilization_percent"])

            logger.info(
                "Memory statistics retrieved successfully", extra={"user_id": user_email, **stats}
            )

            # Track stats API usage
            app_insights.track_custom_event(
                "MemoryStatsAPICall",
                {
                    "user_id": user_email,
                    "total_messages": stats["total_messages"],
                    "memory_utilization": stats["memory_utilization_percent"],
                },
            )

            return response

        except Exception as exc:
            span.record_exception(exc)
            span.set_status(Status(StatusCode.ERROR, str(exc)))

            logger.error(
                "Failed to retrieve memory statistics",
                extra={"user_id": user_email, "error": str(exc), "error_type": type(exc).__name__},
                exc_info=True,
            )

            app_insights.track_exception(exc)

            raise HTTPException(
                status_code=500, detail="Failed to retrieve memory statistics"
            ) from exc


@router.delete("/clear")
async def clear_user_memory(
    current_user: Annotated[dict[str, Any], Depends(get_optional_user)],
    thread_id: str | None = Query(default=None, description="Clear specific thread only"),
) -> JSONResponse:
    """
    Clear user's conversation memory.

    This endpoint allows users to clear their conversation history,
    either entirely or for a specific thread.

    WARNING: This action is irreversible.
    """
    with tracer.start_as_current_span("api.memory.clear") as span:
        user_email = get_user_email(current_user)
        span.set_attribute("user_id", user_email)
        span.set_attribute("thread_specific", thread_id is not None)

        try:
            memory_service = get_memory_service()

            if thread_id:
                # For thread-specific clearing, we'd need to implement this in the memory service
                # For now, return not implemented
                raise HTTPException(
                    status_code=501, detail="Thread-specific memory clearing not yet implemented"
                )
            # Clear all user memory
            store = await memory_service._get_store()
            deleted_count = await store.forget_user(user_email)

            span.set_attribute("deleted_count", deleted_count)

            logger.warning(
                "User memory cleared",
                extra={
                    "user_id": user_email,
                    "deleted_count": deleted_count,
                    "thread_id": thread_id,
                },
            )

            # Track memory clearing in Application Insights
            app_insights.track_custom_event(
                "UserMemoryCleared",
                {
                    "user_id": user_email,
                    "deleted_count": deleted_count,
                    "thread_specific": thread_id is not None,
                },
            )

            return JSONResponse(
                content={
                    "message": "Memory cleared successfully",
                    "deleted_messages": deleted_count,
                    "user_id": user_email,
                }
            )

        except HTTPException:
            raise
        except Exception as exc:
            span.record_exception(exc)
            span.set_status(Status(StatusCode.ERROR, str(exc)))

            logger.error(
                "Failed to clear user memory",
                extra={"user_id": user_email, "error": str(exc), "error_type": type(exc).__name__},
                exc_info=True,
            )

            app_insights.track_exception(exc)

            raise HTTPException(status_code=500, detail="Failed to clear user memory") from exc

File: apps/web/src/lib/memory.ts
/**
 * Memory API client for user conversation management
 * 
 * Provides TypeScript client functions for interacting with the memory service,
 * enabling persistent conversation context and smart AI interactions.
 */

import { API_BASE_URL } from "./api"

export interface ConversationMessage {
    role: "user" | "assistant" | "system" | "tool"
    content: string
    timestamp?: string
    metadata?: Record<string, any>
}

export interface ConversationHistory {
    user_id: string
    messages: ConversationMessage[]
    total_count: number
    thread_id: string | null
    has_more: boolean
}

export interface MessageSearchResult {
    user_id: string
    query: string
    results: ConversationMessage[]
    total_found: number
}

export interface UserMemoryStats {
    user_id: string
    total_messages: number
    recent_user_messages: number
    recent_assistant_messages: number
    memory_utilization_percent: number
    max_memory_limit: number
}

/**
 * Fetch user's conversation history with optional filtering
 */
export async function getConversationHistory(options?: {
    limit?: number
    threadId?: string
    includeMetadata?: boolean
    signal?: AbortSignal
}): Promise<ConversationHistory> {
    const params = new URLSearchParams()
    
    if (options?.limit) params.append('limit', options.limit.toString())
    if (options?.threadId) params.append('thread_id', options.threadId)
    if (options?.includeMetadata) params.append('include_metadata', 'true')
    
    const url = `${API_BASE_URL}/memory/history${params.toString() ? `?${params}` : ''}`
    
    const response = await fetch(url, {
        method: 'GET',
        headers: {
            'Content-Type': 'application/json',
        },
        signal: options?.signal,
    })
    
    if (!response.ok) {
        throw new Error(`Failed to fetch conversation history: ${response.statusText}`)
    }
    
    return response.json()
}

/**
 * Search through user's message history
 */
export async function searchMessages(
    query: string,
    options?: {
        limit?: number
        threadId?: string
        signal?: AbortSignal
    }
): Promise<MessageSearchResult> {
    const response = await fetch(`${API_BASE_URL}/memory/search`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            query,
            limit: options?.limit || 10,
            thread_id: options?.threadId,
        }),
        signal: options?.signal,
    })
    
    if (!response.ok) {
        throw new Error(`Failed to search messages: ${response.statusText}`)
    }
    
    return response.json()
}

/**
 * Get comprehensive memory usage statistics
 */
export async function getMemoryStats(signal?: AbortSignal): Promise<UserMemoryStats> {
    const response = await fetch(`${API_BASE_URL}/memory/stats`, {
        method: 'GET',
        headers: {
            'Content-Type': 'application/json',
        },
        signal,
    })
    
    if (!response.ok) {
        throw new Error(`Failed to fetch memory stats: ${response.statusText}`)
    }
    
    return response.json()
}

/**
 * Clear user's conversation memory
 * WARNING: This action is irreversible
 */
export async function clearMemory(options?: {
    threadId?: string
    signal?: AbortSignal
}): Promise<{ message: string; deleted_messages: number; user_id: string }> {
    const params = new URLSearchParams()
    if (options?.threadId) params.append('thread_id', options.threadId)
    
    const url = `${API_BASE_URL}/memory/clear${params.toString() ? `?${params}` : ''}`
    
    const response = await fetch(url, {
        method: 'DELETE',
        headers: {
            'Content-Type': 'application/json',
        },
        signal: options?.signal,
    })
    
    if (!response.ok) {
        throw new Error(`Failed to clear memory: ${response.statusText}`)
    }
    
    return response.json()
}

/**
 * Load initial conversation context when starting a chat session
 */
export async function loadConversationContext(threadId?: string): Promise<ConversationMessage[]> {
    try {
        const history = await getConversationHistory({
            limit: 10, // Load last 10 messages for context
            threadId,
            includeMetadata: false
        })
        
        console.log(`Loaded ${history.messages.length} messages from conversation history`)
        return history.messages
    } catch (error) {
        console.warn('Failed to load conversation context:', error)
        return []
    }
}

/**
 * Check if user has significant conversation history
 */
export async function hasConversationHistory(): Promise<boolean> {
    try {
        const stats = await getMemoryStats()
        return stats.total_messages > 0
    } catch (error) {
        console.warn('Failed to check conversation history:', error)
        return false
    }
}